### ISSUE TITLE
Issues running in Grails 3.3.6

### ISSUE THREAD

**Issue by paulito415** [bluesliverx/grails-spring-security-oauth2-provider#145]
This issue is a roll-up of issues that were encountered when trying to use the grails-spring-security-oauth2-provider plug-in in Grails 3.3.6. Needless to say, there is already a PR pending for 3.3.x support tracked as issue #141 but that has been around for about a year now. So really not sure when the authors do it. In the mean time, I just wanted to document some of the issues that I've encountered and whether I have found workarounds for them, in case others run into the same issues.

So bottom line up front, the issues I found are summarized below:

1. Wrong package name of FilterRegistration Bean as mentioned in #140 
1. Cannot authenticate with `/oath/token` endpoint (transactionManager issue)
1. For client_credentials login I get an StreamCorruptedException: invalid handle value: 007E0014
1. For some username/password combo, I get an InternalError in ObjectStreamClass.requireInitialized

For the 1st problem, thanks to @sagrawal31 I was able to get past the issue with the FilterRegistrationBean mentioned in #140. I didn't encounter what @piperpilot mentioned because the version I'm using now is actually already 3.0.0-RC2.

For the 2nd problem, this took a longer time because of how Exceptions were wrapped and there was no clear documentation on how I can override Exception handling in Grails. The only thing I had to go on was that the JSON response to the /oauth/token request returned was:

```
{
  "error": "unauthorized", 
  "error_description": "No transactionManager was specified. Using @Transactional or @Rollback requires a valid configured transaction manager. If you are running in a unit test ensure the test has been properly configured and that you run the test suite not an individual test method." 
}
```

Naturally I suspected that this was due to the change in the 'Transactional' package name change according to http://docs.grails.org/latest/guide/upgrading.html. But didn't really know which ones are the offending classes.

Anyway, I was able to look in the logs after turning them on in logback.groovy for `org.springframework.security.oauth2` and `grails.plugin.springsecurity.oauthprovider`, which led me to find out that the the exception was logged in org.springframework.security.authentication.dao.DaoAuthenticationProvider.retrieveUser in the call to loadUserByUsername. During my inspection of sources from the spring-security-core-3.0.0-sources.jar, using "Find in Path" from there for `@Transactional` and found that two classes were using the old package:
- grails.plugin.springsecurity.userdetails.GormUserDetailsService
- grails.plugin.springsecurity.SpringSecurityService.

As an aside, in my investigation I also found that the `@Transactional` annotation was used in at least in 4 classes in grails.plugin.springsecurity.oauthprovider package:

- GormApprovalStoreService
- GormAuthorizationCodeService
- GormClientDetailsService
- GormTokenStoreService

But the only place they were referred to was in grails-plugin.xml which didn't really allow me to override using resources.groovy. 

In any case, after having found GormUserDetailsService and SpringSecurityService as the classes I need to override I overrode them in resources.groovy by copying snippets from the SpringSecurityCoreGrailsPlugin class:

```
  userDetailsService(MyGormUserDetailsService) {
    grailsApplication = grailsApplication
  }

  springSecurityService(MySpringSecurityService) {
    authenticationTrustResolver = ref('authenticationTrustResolver')
    grailsApplication = grailsApplication
    objectDefinitionSource = ref('objectDefinitionSource')
    passwordEncoder = ref('passwordEncoder')
  }
```

For the 3rd problem, I was finally able to override the grailsDispatcherServlet to print the stacktrace when an exception happened, by adding in `resources.groovy`

```
  dispatcherServlet(MyDispatcherServlet)
```

to find the following stacktrace:

```
java.io.StreamCorruptedException: invalid handle value: 007E0014
        at java.io.ObjectInputStream.readHandle(ObjectInputStream.java:1685)
        at java.io.ObjectInputStream.readTypeString(ObjectInputStream.java:1648)
        at java.io.ObjectStreamClass.readNonProxy(ObjectStreamClass.java:791)
        at java.io.ObjectInputStream.readClassDescriptor(ObjectInputStream.java:891)
        at java.io.ObjectInputStream.readNonProxyDesc(ObjectInputStream.java:1855)
        at java.io.ObjectInputStream.readClassDesc(ObjectInputStream.java:1749)
        at java.io.ObjectInputStream.readOrdinaryObject(ObjectInputStream.java:2040)
        at java.io.ObjectInputStream.readObject0(ObjectInputStream.java:1571)
        at java.io.ObjectInputStream.defaultReadFields(ObjectInputStream.java:2285)
        at java.io.ObjectInputStream.readSerialData(ObjectInputStream.java:2209)
        at java.io.ObjectInputStream.readOrdinaryObject(ObjectInputStream.java:2067)
        at java.io.ObjectInputStream.readObject0(ObjectInputStream.java:1571)
        at java.io.ObjectInputStream.readObject(ObjectInputStream.java:431)
        at java_io_ObjectInput$readObject.call(Unknown Source)
        at grails.plugin.springsecurity.oauthprovider.serialization.DefaultOAuth2AuthenticationSerializer$_deserialize_closure1.doCall(DefaultOAuth2AuthenticationSerializer.groovy:17)

```

For the 4th problem, again with the exception handling override, I was able to find the problem happened at the same place as in the 3rd problem:

```
Caused by: java.lang.InternalError: Unexpected call when not initialized
        at java.io.ObjectStreamClass.requireInitialized(ObjectStreamClass.java:851)
        at java.io.ObjectStreamClass.checkDeserialize(ObjectStreamClass.java:860)
        at java.io.ObjectInputStream.readOrdinaryObject(ObjectInputStream.java:2041)
        at java.io.ObjectInputStream.readObject0(ObjectInputStream.java:1571)
        at java.io.ObjectInputStream.skipCustomData(ObjectInputStream.java:2251)
        at java.io.ObjectInputStream.readSerialData(ObjectInputStream.java:2213)
        at java.io.ObjectInputStream.readOrdinaryObject(ObjectInputStream.java:2067)
        at java.io.ObjectInputStream.readObject0(ObjectInputStream.java:1571)
        at java.io.ObjectInputStream.defaultReadFields(ObjectInputStream.java:2285)
        at java.io.ObjectInputStream.readSerialData(ObjectInputStream.java:2209)
        at java.io.ObjectInputStream.readOrdinaryObject(ObjectInputStream.java:2067)
        at java.io.ObjectInputStream.readObject0(ObjectInputStream.java:1571)
        at java.io.ObjectInputStream.defaultReadFields(ObjectInputStream.java:2285)
        at java.io.ObjectInputStream.readSerialData(ObjectInputStream.java:2209)
        at java.io.ObjectInputStream.readOrdinaryObject(ObjectInputStream.java:2067)
        at java.io.ObjectInputStream.readObject0(ObjectInputStream.java:1571)
        at java.io.ObjectInputStream.readObject(ObjectInputStream.java:431)
        at java_io_ObjectInput$readObject.call(Unknown Source)
        at grails.plugin.springsecurity.oauthprovider.serialization.DefaultOAuth2AuthenticationSerializer$_deserialize_closure1.doCall(DefaultOAuth2AuthenticationSerializer.groovy:17)

```

both the 3rd and 4th problems look a bit too low level to debug. I may try to fix the `@Transactional` problem in the 4 classes I found before trying anything else.

**Comment by paulito415** [comment]
So on a leap of faith, I discovered that the magic that is `resources.groovy` is indeed pretty magical, for all 4 classes I was able to override them using resources.groovy declarations, just following the proper naming conventions, as below:

```
  gormApprovalStoreService(AcGormApprovalStoreService){
    grailsApplication = grailsApplication
  }

  gormAuthorizationCodeService(AcGormAuthorizationCodeService) {
    grailsApplication = grailsApplication
    oauth2AuthenticationSerializer = ref('oauth2AuthenticationSerializer')
  }

  gormClientDetailsService(AcGormClientDetailsService) {
    grailsApplication = grailsApplication
    clientAdditionalInformationSerializer = ref('clientAdditionalInformationSerializer')
  }

  gormTokenStoreService(AcGormTokenStoreService) {
    grailsApplication = grailsApplication
    oauth2AuthenticationSerializer = ref('oauth2AuthenticationSerializer')
    authenticationKeyGenerator = ref('authenticationKeyGenerator')
    accessTokenAdditionalInformationSerializer = ref('accessTokenAdditionalInformationSerializer')
    accessTokenScopeSerializer = ref('accessTokenScopeSerializer')
  }
```

But alas that didn't solve my issue with the deserialization. So I went a bit further and in my override of the GormTokenStoreService (MyGormTokenStoreService), I printed out what was being deserialized and it seems it was the byte[] that is stored in the `authentication` column of my AccessToken class, to be deserialized into an `OAuth2Authentication` object.

What was very odd was that this was only happening for a specific user in my system, but after I deleted that user and restarted, the same thing happened. But given that this has to do with the AccessToken class, I looked into the access_token table and I noticed that there were a couple of existing rows for my username / password in the access_token table. What I don't understand is why my deletion of the User didn't cause a foreign key constraint violation. 

Thinking that their existence caused unnecessary deserialization causing the issue I saw, and although I personally do not know the policy with which the plugin stores access token information into the database, I guessed that the rows can be recreated by `AccessTokenStoreService` code after deletion, so I went ahead and deleted the rows.

This turned out to be the solution I needed to solve problem 4 above. So now 3 down 1 to go.

**Comment by paulito415** [comment]
Turns out that for problem 3 also, all I needed to do is to delete rows in the access_token table where the username is null (apparently client_credentials grant also uses this same table). And so now all 4 issues are resolved.

**Comment by ChiranjibKarn** [comment]
I am also stuck on this (client credential grant type), would be helpful if you provide pull request.

**Comment by paulito415** [comment]
@ChiranjibKarn my advice is to delete all rows in your access_token table if you can (at least when your system is offline) so that they get refreshed when sought thereafter.

**Comment by bluesliverx** [comment]
My apologies, others have been managing this repo for quite a long time now.  I'm actually in the process of upgrading my (very old, on 2.3.11) grails app to 3.3.8, so I will run into this very shortly.  I will be merging these patches shortly and publishing a new version of the plugin, unless there is already a fork that has been "officially" released.  I'm planning to do this over the weekend most likely, maybe sooner if I can get to it.

**Comment by bluesliverx** [comment]
@paulito415 It's been awhile and I'm not sure of the status of this issue anymore. Has it been resolved with the 3.1.0-RC1 release or are there still problems to address? If there is still an issue, please reopen and let me know so I can migrate the issue to the new project.
